File: src/scripts/backfill-duckdb.ts
#!/usr/bin/env bun
/**
 * Backfill DuckDB/MotherDuck from ClickHouse
 *
 * Pulls ALL historical data from ClickHouse ccusage_events table
 * and writes to DuckDB (local or MotherDuck).
 */

import * as dotenv from 'dotenv';
dotenv.config();

import { CHClient } from '../database/client.js';
import { ClickHouseConfig } from '../config/clickhouse.js';
import { DuckDBSink } from '../sinks/duckdb.js';
import type { EventsSnapshotData } from '../pipeline/types.js';

const args = process.argv.slice(2);
const verbose = args.includes('--verbose') || args.includes('-v');
const duckdbPath = args.find(a => a.startsWith('--path='))?.split('=')[1] || process.env.DUCKDB_PATH || 'md:ccusage';

console.log(`Backfill: ClickHouse ccusage_events → ${duckdbPath}`);

const chConfig = ClickHouseConfig.fromEnv();
const client = new CHClient(chConfig);

// Read from ccusage_events table
const rows = await client.query<Record<string, unknown>>('SELECT * FROM ccusage_events');
await client.close();

console.log(`  ccusage_events: ${rows.length} rows`);

const data: EventsSnapshotData = { events: rows };

console.log(`\nWriting ${rows.length} rows to ${duckdbPath}...`);
const sink = new DuckDBSink({ dbPath: duckdbPath });
await sink.connect();
const result = await sink.write(data);
await sink.close();

console.log(`Done: ${Object.values(result.rowsWritten).reduce((a, b) => a + b, 0)} rows in ${result.durationMs}ms`);
for (const [table, count] of Object.entries(result.rowsWritten)) {
  console.log(`  ${table}: ${count}`);
}

File: src/script/backfill_duckdb.rs
use crate::config::ClickHouseConfig;
use crate::model::EventsSnapshotData;
use crate::sink::duckdb::DuckDbSink;
use crate::sink::DataSink;
use crate::util::timer::CommandTimeout;
use reqwest::Client;
use std::env;
use std::time::Duration;

/// Backfill DuckDB/MotherDuck from ClickHouse.
///
/// Reads all rows from ClickHouse `ccusage_events` and writes them
/// to the configured DuckDB path.
pub fn run() -> anyhow::Result<()> {
    dotenvy::dotenv().ok();

    let args: Vec<String> = env::args().collect();
    let verbose = args.contains(&"--verbose".into()) || args.contains(&"-v".into());

    let path = env::var("DUCKDB_PATH")
        .or_else(|_| {
            args.iter()
                .find(|a| a.starts_with("--path="))
                .map(|a| a.split('=').nth(1).unwrap_or("md:ccusage"))
        })
        .unwrap_or_else(|_| "md:ccusage".into());

    println!("Backfill: ClickHouse ccusage_events → {}", path);

    let cfg = ClickHouseConfig::from_env();
    let timeout = CommandTimeout::from_ms(
        env::var("IMPORT_COMMAND_TIMEOUT_MS")
            .ok()
            .and_then(|v| v.parse().ok()),
    )
    .duration()
    .unwrap_or_else(|| Duration::from_secs(30));

    let client = Client::builder()
        .timeout(timeout)
        .build()
        .expect("build reqwest client");

    let url = format!(
        "{}://{}:{}/?query={}",
        cfg.protocol,
        cfg.host,
        cfg.port,
        crate::sink::clickhouse::percent_encode("SELECT * FROM ccusage_events")
    );

    let rows: Vec<serde_json::Value> = client
        .get(&url)
        .basic_auth(
            cfg.user.clone(),
            if cfg.password.is_empty() {
                None
            } else {
                Some(cfg.password.clone())
            },
        )
        .send()
        .and_then(|r| r.error_for_status())
        .expect("fetch from ClickHouse")
        .json()
        .expect("parse ClickHouse response");

    println!("  ccusage_events: {} rows", rows.len());

    let events = rows
        .into_iter()
        .map(|v| v)
        .collect::<Vec<serde_json::Value>>();

    println!("\nWriting {} rows to {}...", events.len(), path);

    let mut sink = DuckDbSink::new(path);
    sink.connect().expect("connect duckdb");

    let start = std::time::Instant::now();
    let result = sink
        .write(EventsSnapshotData { events })
        .expect("write to duckdb");
    sink.close().expect("close duckdb");

    let total = result.rows_written.values().copied::<u64>().sum::<u64>();
    println!(
        "Done: {} rows in {}ms",
        total,
        start.elapsed().as_millis()
    );
    for (table, count) in result.rows_written {
        println!("  {}: {}", table, count);
    }

    Ok(())
}
